Install current release then upgrade

newtewt opened the pull request:
# Description

On normal builds not much changes here. On tag builds, we first launch horti and install the latest medic channel release.  After it starts and bootstraps everything. It shuts down that container and starts a new horti instance with the beta tag build. 

Tag Build
https://github.com/medic/cht-core/actions/runs/1237828216

Branch Build
https://github.com/medic/cht-core/actions/runs/1237779865


Automates: https://github.com/medic/cht-release-testing/issues/55

# Code review checklist
<!-- Remove or comment out any items that do not apply to this PR; in the remaining boxes, replace the [ ] with [x]. -->
- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

dianabarsan commented:
aw man ...I think your merge ended up showing all changes from master in the diff :(

dianabarsan commented on scripts/ci/horti.env:
```diff
@@ -0,0 +1,5 @@
+MARKET_URL_READ=https://staging.dev.medicmobile.org
+BUILDS_SERVER=_couch/builds_testing
+BUILD_NUMBER=1221218955
+NODE=14.x
+TAG=3.11.1-beta.1
```
Are all these values placeholders?

newtewt replied:
Yeah they are for local testing. So you can do `docker-compose -f scripts/ci/horti_compose.yml --env-file scripts/ci/horti.env up ` for testing locally. I was conflicted on whether or not to commit this but felt it might help when troubleshooting.

dianabarsan replied:
Can we please then use a Node version that we support? :) 
Also, please add some comments about what these are.

dianabarsan commented on scripts/ci/horti_entry.sh:
```diff
@@ -30,8 +30,18 @@ echo Creating logs dir
+fi
+
+echo 'Horti Env Vars'
+echo Build server url $HORTI_BUILDS_SERVER
```
Looking at a line above: 
```
DEFAULT_BUILDS_URL=https://staging.dev.medicmobile.org/_couch/builds
...
HORTI_BUILDS_SERVER=$DEFAULT_BUILDS_URL
```

Just trying to make sure we won't accidentally log passwords and such, is this value always safe to echo?

newtewt replied:
When using horti to install anything. We are using a separate READ_URL that should never have the password. The only one to have the password now is on publish.  The only thing changing is the endpoint.

dianabarsan replied:
Excellent! Thanks for clarifying!

dianabarsan commented on .github/workflows/build.yml:
```diff
@@ -133,11 +136,23 @@ jobs:
     - name: npm CI
       run: npm ci
-    - name: horti setup
+    - name: Horti Setup
```
Just a nitpick. 

For me, it would be easier to understand if the extra horti step, that would only execute if it's a tag, was better separated from the other one. 

I would change the "RESTART" param to a "INSTALL_LATEST" param and pass it to this extra step, and change the condition in the script to check for `IS_TAG` && `INSTALL_LATEST` to change the builds server. 

I'd also make this step be executed before `Horti Setup` and give it a suggestive name like `Install latest`.

dianabarsan reviewed (commented):
Nice quick work on this. 
I tried to minify all the covid-19 config and add some comments, but I may have missed some files.

newtewt commented on .github/workflows/build.yml:
```diff
@@ -133,11 +136,23 @@ jobs:
+    # Sleeping here to ensure the previous run gets through all processing api does on first startup.
+    - name: Horti Restart
+      run: |
+        sleep 120
```
I agree, is there anything that signifies API is done bootstrapping(xforms, migrations, etc...)?

dianabarsan replied:
We only start the express server once all bootstrap steps are done: 
https://github.com/medic/cht-core/blob/master/api/server.js#L86

So API responding to requests means that it's bootstrapped successfully.

newtewt replied:
ha, that makes it easy.

newtewt commented:
> aw man ...I think your merge ended up showing all changes from master in the diff :(

I've updated to only show what I have changed but I believe your comments were all in my changes anyways :-D

newtewt commented:
@dianabarsan  I ran into this during the tag build. Without a fix in horti. Is there a way I can avoid this via script/curl/couchdb req?

https://github.com/medic/horticulturalist/issues/59

dianabarsan commented:
> @dianabarsan I ran into this during the tag build. Without a fix in horti. Is there a way I can avoid this via script/curl/couchdb req?
> [medic/horticulturalist#59](https://github.com/medic/horticulturalist/issues/59)

No, sadly. The only way to avoid this is to make sure nothing else touches the `horti-upgrade` doc while horti is installing.

dianabarsan reviewed (approved):
Nice work! 

I wonder why the push build ran twice
![Screenshot_20210917_152222](https://user-images.githubusercontent.com/35681649/133781901-ae9f3c08-1710-437e-a3a3-8665c5ff89e3.png)

Is your local PC time correct?

newtewt commented:
The last one is the tag build. Didn't realize it would associate to the pull request but that is neat.
